docs/proposals/network-policy.md

```diff
@@ -0,0 +1,321 @@
+  // traffic matches at least one NetworkPolicyPeer in the From list. 
+  From []NetworkPolicyPeer `json:"from,omitempty"`
+}
+
+type NetworkPolicyPort struct {
+  // The protocol (TCP or UDP) which traffic must match.
+  // If not defined, this field defaults to TCP. 
+  Protocol api.Protocol `json:"protocol"`
+
+  // If specified, the port on the given protocol.  This can 
+  // either be a numerical or named port.  If this field is nil,
+  // this NetworkPolicyPort matches all port names and numbers.
+  // If non-nil, only traffic on the specified protocol AND port
+  // will be matched by this NetworkPolicyPort.
+  Port *intstr.IntOrString `json:"port,omitempty"`
+}
+
+type NetworkPolicyPeer struct {
+  // If 'Namespaces' is defined, 'Pods' must not be.
+  // This is a label selector which selects Pods in this namespace.
+  // This NetworkPolicyPeer matches any pods selected by this selector.
+  // This field follows standard unversioned.LabelSelector semantics.
+  // If nil, this selector selects no pods.
+  // If non-nil but empty, this selector selects all pods in this namespace.
+  Pods *unversioned.LabelSelector `json:"pods,omitempty"`
+
+  // If 'Pods' is defined, 'Namespaces' must not be.
+  // Selects Kubernetes Namespaces.  This NetworkPolicyPeer matches 
+  // all pods in all namespaces selected by this label selector. 
+  // This field follows standard unversioned.LabelSelector semantics.
+  // If nil, this selector selects no namespaces.
+  // If non-nil but empty, this selector selects all namespaces.
+  Namespaces *unversioned.LabelSelector `json:"namespaces,omitempty"`
+}
+```
+
+### Behavior
+
+The following pseudo-code attempts to define when traffic is allowed to a given pod when using this API.
+
+```python
+def is_traffic_allowed(traffic, pod):
+  """
+  Returns True if traffic is allowed to this pod, False otherwise.
+  """
+  if not namespace.networkPolicy.ingress.isolation:
+    # If ingress isolation is disabled on the Namespace, all traffic is allowed.
+    return True 
+  elif traffic.source == pod.node.kubelet:
+    # Traffic is from kubelet health checks.
+    return True
+  else:
+    # If namespace ingress isolation is enabled, only allow traffic 
+    # that matches a network policy which selects this pod.
+    for network_policy in network_policies(pod.namespace):
+      if not network_policy.podSelector.selects(pod):
+        # This policy doesn't select this pod. Try the next one. 
+        continue
+
+      # A null ingress list means allow all traffic.
+      if network_policy.ingress == null:
+        return True
+
+      # This policy selects this pod.  Check each ingress rule 
+      # defined on this policy to see if it allows the traffic.
+      # If at least one does, then the traffic is allowed.
+      for ingress_rule in network_policy.ingress:
+        if ingress_rule.matches(traffic): 
```

Comment:
I don't think it should.  The Golang representation is meant to be the source of truth for how this API works.  This code is meant as an example to show how the API behaves at a high-level.

If a reader is interested in how matching for ingress rules work, then she should look to the Go structs and comments above.